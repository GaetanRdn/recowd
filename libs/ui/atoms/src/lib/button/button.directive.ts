import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
import { CoerceBoolean } from '@recowd/utility-types';

@Directive({
  selector:
    'button[rcButtonIcon], a[rcButtonIcon], button[rcButton], a[rcButton]',
  standalone: true,
})
export class ButtonDirective {
  @Input() @CoerceBoolean() @HostBinding('class.rc-disabled') disabled = false;

  constructor(private readonly _elementRef: ElementRef<HTMLElement>) {}

  @HostBinding('class.rc-button-icon')
  get isButtonIcon(): boolean {
    return this.hostElement().getAttributeNames().includes('rcbuttonicon');
  }

  @HostBinding('class.rc-button')
  get isButton(): boolean {
    return this.hostElement().getAttributeNames().includes('rcbutton');
  }

  private hostElement(): HTMLElement {
    return this._elementRef.nativeElement;
  }
}
