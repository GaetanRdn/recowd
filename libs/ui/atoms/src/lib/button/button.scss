@mixin rc-button($config) {
  $colors: map-get($config, colors);
  $primary: map-get($colors, primary);

  button {
    appearance: none;
    border: none;
  }

  a {
    text-decoration: none;
  }

  .rc-button-icon {
    color: map-get($primary, color);
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;
    padding: 0.4rem;
    background-color: transparent;
  }

  .rc-button-icon:is(.rc-disabled) > * {
    color: map-get(map-get($colors, disabled), color);
  }
}
