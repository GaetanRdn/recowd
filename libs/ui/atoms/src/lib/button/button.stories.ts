import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { ButtonDirective, ButtonModule } from './button.directive';
import { MaterialIconModule } from '../material-icon/material-icon.component';
import { BADGE } from '@geometricpanda/storybook-addon-badges';

export default {
  title: 'atoms/buttons',
  component: ButtonDirective,
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'button.directive',
  },
  decorators: [
    moduleMetadata({
      imports: [ButtonModule, MaterialIconModule],
    }),
  ],
} as Meta<ButtonDirective>;

const templateForButtonIcon: Story<ButtonDirective> = (
  args: ButtonDirective
) => ({
  props: {
    disabled: args.disabled,
  },
  template: `<button rcButtonIcon [disabled]="disabled"><rc-material-icon>settings</rc-material-icon></button>`,
});

export const ButtonIcon = templateForButtonIcon.bind({});
ButtonIcon.args = {
  disabled: false,
};

const templateForLinkIcon: Story<ButtonDirective> = (
  args: ButtonDirective
) => ({
  props: {
    disabled: args.disabled,
  },
  template: `<a rcButtonIcon [disabled]="disabled"><rc-material-icon>home</rc-material-icon></a>`,
});

export const LinkIcon = templateForLinkIcon.bind({});
LinkIcon.args = {
  ...ButtonIcon.args,
};
