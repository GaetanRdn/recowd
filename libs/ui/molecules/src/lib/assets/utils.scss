@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
