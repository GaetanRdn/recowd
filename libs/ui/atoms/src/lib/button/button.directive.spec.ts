import { Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { TestBed, waitForAsync } from '@angular/core/testing';
import { ButtonDirective } from './button.directive';
import { TemplateLookup } from '@recowd/test/utils';

describe('ButtonDirective', () => {
  beforeEach(waitForAsync(() =>
    TestBed.configureTestingModule({
      declarations: [
        ButtonDirective,
        ButtonIconHostComponent,
        LinkIconHostComponent,
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents()));

  describe('ButtonIcon', () => {
    let templateLookup: TemplateLookup<ButtonIconHostComponent>;

    beforeEach(
      () => (templateLookup = new TemplateLookup(ButtonIconHostComponent))
    );

    it('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be disabled', () => {
      // WHEN
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe('LinkIcon', () => {
    let templateLookup: TemplateLookup<LinkIconHostComponent>;

    beforeEach(
      () => (templateLookup = new TemplateLookup(LinkIconHostComponent))
    );

    it('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    it('should be disabled', () => {
      // WHEN
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

@Component({
  template: ` <button rcButtonIcon [disabled]="disabled">
    <rc-material-icon>home</rc-material-icon>
  </button>`,
})
class ButtonIconHostComponent {
  public disabled = false;
}

@Component({
  template: `<a rcButtonIcon [disabled]="disabled">
    <rc-material-icon>home</rc-material-icon>
  </a>`,
})
class LinkIconHostComponent extends ButtonIconHostComponent {}
