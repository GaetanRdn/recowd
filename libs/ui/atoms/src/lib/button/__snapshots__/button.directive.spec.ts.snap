// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDirective ButtonIcon should be disabled 1`] = `
<button
  class="rc-button-icon rc-disabled"
  rcbuttonicon=""
>
  <rc-material-icon>
    home
  </rc-material-icon>
</button>
`;

exports[`ButtonDirective ButtonIcon should create 1`] = `
<button
  class="rc-button-icon"
  rcbuttonicon=""
>
  <rc-material-icon>
    home
  </rc-material-icon>
</button>
`;

exports[`ButtonDirective LinkIcon should be disabled 1`] = `
<a
  class="rc-button-icon rc-disabled"
  rcbuttonicon=""
>
  <rc-material-icon>
    home
  </rc-material-icon>
</a>
`;

exports[`ButtonDirective LinkIcon should create 1`] = `
<a
  class="rc-button-icon"
  rcbuttonicon=""
>
  <rc-material-icon>
    home
  </rc-material-icon>
</a>
`;
